### DISCUSSION TITLE
arm64 and armv7 support

### DISCUSSION THREAD

**Discussion by LoganJFisher** [uniuuu/zotprime#35] [Ideas]
Please consider adding support for arm64 and armv7. I would like to use a Home Assistant add-on built off of this repository, but being limited to amd64 is inconvenient for Home Assistant as many users are on arm-based systems.

This change would allow for the creation of a broadly usable Home Assistant add-on, which would in turn make running zotprime significantly more accessible.

Please see this discussion for further details: https://github.com/alexbelgium/hassio-addons/issues/1227

**Comment by uniuuu** [comment]
Hi @LoganJFisher 

Thank you.
I'll have some free time on weekend and review it.

**Comment by uniuuu** [comment]
Hi @LoganJFisher 
There is a chance that it will run on ARM. However might be that not all components will run without issues. 
Would you be able to provide a tested/verified on ARM emulator i.e. qemu-system-arm virtual machine image with Debian/Ubuntu?

**Comment by LoganJFisher** [comment]
Sorry, but that is FAR beyond my skillset.

**Comment by uniuuu** [comment]
Currently it's in progress to test all Zotero Server components in ARM emulation. Cannot tell ETA for it, probably next weekend.
For Zotero Client there is no yet ARM available from official page. But there is Arch Linux package that can be converted to DEB package as per https://forums.zotero.org/discussion/95586/zotero-on-arm64-linux-or-chromeos-arm

**Comment by uniuuu** [comment]
The latest Zotero Client v 7 added ARM64 support https://github.com/uniuuu/zotprime/discussions/30

**Comment by uniuuu** [comment]
@LoganJFisher I do not have Windows with ARM around. 
If you have one could you try to build a new v 7 client there ?

**Comment by uniuuu** [comment]
Moving this topic to discussions

**Comment by LoganJFisher** [comment]
I don't have Windows with ARM either, and building a client is beyond my skillset anyway.

My ARM devices are a Raspberry Pi which I use to run Home Assistant (and the hope for [alexbelgium](https://github.com/alexbelgium/hassio-addons/issues/1227) to develop a Zotero addon was the initial goal of this chain of inquires), and a Synology NAS.

**Reply by uniuuu** [reply]
@LoganJFisher We'll get it work. I will make emulation of ARM64 and once we have all containers running then you can push it further to test on a real hardware for hass.
I understood you're not looking into Zotero client running on ARM architecture but Zotero data server only.

**Reply by uniuuu** [reply]
Hi @LoganJFisher 
I have progress here however one of components is failing to build for ARM64 due to old dependencies when run newer node alpine image and old working image is not available for ARM64. I plan to move it from node alpine image to debian image to test if we can get all these issues settled.

**Comment by uniuuu** [comment]
Hi @LoganJFisher 
Good day.
It's been released. Please clone latest code and pull latest images they're supporting ARM64 already. 
However ARM7 support blocked by elasticseach image.

Other services support more:
- MariaDB: amd64, arm64, ppc64le, s390x
- Redis: amd64, arm64, armv7, armv6, 386, ppc64le, s390x, mips64le
- Alpine/Node: amd64, arm64, armv7, armv6, ppc64le, s390x, 386

**Comment by haecks** [comment]
Hi all! 
As i recently adopted this project on an Nvidia Spark (aarch64) i observed several issues to be worked around, not all related to the projects docker-images, i like to share:

1. The `zotprime-phpmyadmin`-image referenced by `image: "uniuu/zotprime-phpmyadmin:${VER}"` - as well as the original image `phpmyadmin/phpmyadmin:latest` - holds an incompatible `bash`-binary:
`bash: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=8b362acb36d9f6e7103ef8433c2141265908ac8f, for GNU/Linux 3.2.0, stripped`
So it was necessary to switch to `image: "arm64v8/phpmyadmin:latest"` instead (could be achieved e.g. by additional file `docker-compose.override.yml`like
```
services:
  zotprime-phpmyadmin:
    image: "arm64v8/phpmyadmin:latest"
```
). 

2. The `zotprime-localstack`-base-image needed to be replaced as well on my system switching from `image: "uniuu/zotprime-localstack:${VER}"` to `image: "localstack/localstack"`, This sadly implies necessity for adopting all occurences of `http://zotprime-localstack:4575`  to `http://zotprime-localstack:4566` in `/stack/init/auto-init.sh`, as the port is then changed. Another option might be adding a port-mapping like 
```
  ports:
      - "4575:4566"
```
 in the `docker-compose.yml`'s defintion of the `zotprime-localstack`-service (not yet checked).

Maybe this is helpful for somebody.

**Reply by uniuuu** [reply]
@haecks Thank you for sharing it. I'll check it all. 
I have plans to replace localstack with nats as that is very old image and not a production kind.

**Reply by uniuuu** [reply]
@haecks I've checked the image that is used https://hub.docker.com/_/phpmyadmin/tags - supports it. 
I pushed fix to image build CI. Both localstack and phpmyadmin new images should be fixed already. Please check.
